File: packages/translation/ios/Plugin/Translation.swift
import Foundation
import MLKitTranslate

@objc public class Translation: NSObject {
    private let plugin: TranslationPlugin

    private var downloadQueue: [String: ((Error?) -> Void)] = {
        return [String: ((Error?) -> Void)]()
    }()

    init(plugin: TranslationPlugin) {
        self.plugin = plugin
    }

    @objc public func deleteDownloadedModel(language: TranslateLanguage, completion: @escaping (Error?) -> Void) {
        let model = TranslateRemoteModel.translateRemoteModel(language: language)
        ModelManager.modelManager().deleteDownloadedModel(model) { error in
            completion(error)
        }
    }

    @objc public func downloadModel(language: TranslateLanguage, completion: @escaping (Error?) -> Void) {
        let model = TranslateRemoteModel.translateRemoteModel(language: language)
        let conditions = ModelDownloadConditions(
            allowsCellularAccess: true,
            allowsBackgroundDownloading: true
        )
        ModelManager.modelManager().download(model, conditions: conditions)
        self.downloadQueue[language.rawValue] = completion
    }

    @objc public func getDownloadedModels() -> Set<TranslateRemoteModel> {
        return ModelManager.modelManager().downloadedTranslateModels
    }

    @objc public func translate(text: String, sourceLanguage: TranslateLanguage, targetLanguage: TranslateLanguage, completion: @escaping (String?, Error?) -> Void) {
        let options = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
        let translator = Translator.translator(options: options)
        let conditions = ModelDownloadConditions(
            allowsCellularAccess: true,
            allowsBackgroundDownloading: true
        )
        translator.downloadModelIfNeeded(with: conditions) { error in
            if let error = error {
                completion(nil, error)
                return
            }

            translator.translate(text) { translatedText, error in
                completion(translatedText, error)
            }
        }
    }

    public func handleModelDownloadDidSucceed(model: TranslateRemoteModel) {
        if let completion = self.downloadQueue[model.language.rawValue] {
            completion(nil)
        }
    }

    public func handleModelDownloadDidFail(model: TranslateRemoteModel) {
        if let completion = self.downloadQueue[model.language.rawValue] {
            completion(plugin.errorDownloadModelFailed)
        }
    }
}

extension String: LocalizedError {
    public var localizedDescription: String? { return self }
}

File: packages/translation/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/translation/Translation.java
package io.capawesome.capacitorjs.plugins.mlkit.translation;

import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.common.model.RemoteModelManager;
import com.google.mlkit.nl.translate.TranslateRemoteModel;
import com.google.mlkit.nl.translate.Translator;
import com.google.mlkit.nl.translate.TranslatorOptions;
import java.util.Set;

public class Translation {

    private RemoteModelManager modelManager;

    public Translation() {
        this.modelManager = RemoteModelManager.getInstance();
    }

    public void deleteDownloadedModel(String language, final DeleteDownloadedModelResultCallback resultCallback) {
        TranslateRemoteModel frenchModel = new TranslateRemoteModel.Builder(language).build();
        modelManager
            .deleteDownloadedModel(frenchModel)
            .addOnSuccessListener(
                (OnSuccessListener) result -> {
                    resultCallback.success();
                }
            )
            .addOnFailureListener(exception -> {
                resultCallback.error(exception);
            });
    }

    public void downloadModel(String language, final DownloadModelResultCallback resultCallback) {
        TranslateRemoteModel frenchModel = new TranslateRemoteModel.Builder(language).build();
        DownloadConditions conditions = new DownloadConditions.Builder().requireWifi().build();
        modelManager
            .download(frenchModel, conditions)
            .addOnSuccessListener(
                (OnSuccessListener) result -> {
                    resultCallback.success();
                }
            )
            .addOnFailureListener(exception -> {
                resultCallback.error(exception);
            });
    }

    public void getDownloadedModels(final GetDownloadedModelsResultCallback resultCallback) {
        modelManager
            .getDownloadedModels(TranslateRemoteModel.class)
            .addOnSuccessListener(
                (OnSuccessListener<Set<TranslateRemoteModel>>) models -> {
                    resultCallback.success(models);
                }
            )
            .addOnFailureListener(exception -> {
                resultCallback.error(exception);
            });
    }

    public void translate(String text, String sourceLanguage, String targetLanguage, final TranslateResultCallback resultCallback) {
        TranslatorOptions options = new TranslatorOptions.Builder()
            .setSourceLanguage(sourceLanguage)
            .setTargetLanguage(targetLanguage)
            .build();
        final Translator translator = com.google.mlkit.nl.translate.Translation.getClient(options);
        DownloadConditions conditions = new DownloadConditions.Builder().requireWifi().build();
        translator
            .downloadModelIfNeeded(conditions)
            .addOnSuccessListener(
                (OnSuccessListener) downloadModelIfNeededResult -> {
                    translator
                        .translate(text)
                        .addOnSuccessListener(
                            (OnSuccessListener<String>) translateResult -> {
                                resultCallback.success(translateResult);
                                translator.close();
                            }
                        )
                        .addOnFailureListener(exception -> {
                            resultCallback.error(exception);
                            translator.close();
                        });
                }
            )
            .addOnFailureListener(exception -> {
                resultCallback.error(exception);
                translator.close();
            });
    }
}
